tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift

```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
       return
     }
 
-    let hSnapshot: HANDLE =
-        CreateToolhelp32Snapshot(DWORD(TH32CS_SNAPHEAPLIST), dwProcessId)
-    if hSnapshot == INVALID_HANDLE_VALUE {
-      // FIXME(compnerd) log error
+    // We use a shared memory and two event objects to send heap entries data
+    // from the remote process to this process. A high-level structure looks
+    // like below:
+    //
+    // Swift inspect (this process):
+    //
+    // Setup the shared memory and event objects
+    // Create a remote thread to invoke the heap walk on the remote process
+    // Loop {
+    //   Wait on ReadEvent to wait for heap entries in the shared memory
+    //   If no entries, break
+    //   Inspect and dump heap entries from the shared memory
+    //   Notify (SetEvent) on WriteEvent to have more heap entries written
+    // }
+    //
+    // Remote process:
+    //
+    // Open the shared memory and event objects
+    // Heap walk loop {
+    //   Write heap entries in the shared memory until full or done
+    //   Notify (SetEvent) ReadEvent to have them read
+    //   Wait on WriteEvent until they are read
+    // }
+    //
+
+    // Exclude the self-inspect case. We use IPC + HeapWalk in the remote
+    // process, which doesn't work on itself.
+    if dwProcessId == GetCurrentProcessId() {
+      print("Cannot inspect the process itself")
       return
     }
-    defer { CloseHandle(hSnapshot) }
 
-    var heap: HEAPLIST32 = HEAPLIST32()
-    heap.dwSize = SIZE_T(MemoryLayout<HEAPLIST32>.size)
-    if !Heap32ListFirst(hSnapshot, &heap) {
-      // FIXME(compnerd) log error
+    // The size of the shared memory buffer and the names of shared
+    // memory and event objects
+    let bufSize = Int(BUF_SIZE)
+    let sharedMemoryName = SHARED_MEM_NAME_PREFIX + "-" + String(dwProcessId)
+    let readEventName = READ_EVENT_NAME_PREFIX + "-" + String(dwProcessId)
+    let writeEventName = WRITE_EVENT_NAME_PREFIX + "-" + String(dwProcessId)
+    let waitTimeoutMs = DWORD(WAIT_TIMEOUT_MS)
+
+    // Set up the shared memory
+    let hMapFile = CreateFileMappingA(INVALID_HANDLE_VALUE,
+                                      UnsafeMutablePointer(bitPattern: 0),
+                                      DWORD(PAGE_READWRITE),
+                                      0,
+                                      DWORD(bufSize),
+                                      sharedMemoryName);
+    if hMapFile == HANDLE(bitPattern: 0) {
+      print("CreateFileMapping failed \(GetLastError())")
+      return
+    }
+    if GetLastError() == ERROR_ALREADY_EXISTS {
+      print("FileMapping in use. Try again in a while.")
+      CloseHandle(hMapFile)
+      return
+    }
+    defer { CloseHandle(hMapFile) }
```

Comment:
Should this be hoisted over the `GetLastError`?

Reply:
Done